8248499: Add methods to allocate off heap arrays from Java arrays

mcimadamore opened the pull request:
This patch adds methods to NativeScope which allow clients to allocate off-heap arrays from Java on-heap arrays.

While in most cases the code for copying is straightforward, we have to take into account endianness mismatches (in case the user wants to store the array elements off-heap in an endianness that doesn't match that of the platform).

To allow for that, I've added a new implementation method in AbstractMemorySegmentImpl, which is the same as `copyFrom`, but adds the byte swap (a wrapper around `Unsafe::copySwapMemory`).
<!-- Anything below this marker will be automatically updated, please do not edit manually! -->
---------
### Progress
- [x] Change must not contain extraneous whitespace
- [x] Change must be properly reviewed

### Issue
 * [JDK-8248499](https://bugs.openjdk.java.net/browse/JDK-8248499): Add methods to allocate off heap arrays from Java arrays


### Reviewers
 * Athijegannathan Sundararajan ([sundar](@sundararajana) - Committer) ⚠️ Review applies to ac6e37dc7c3eb91f4640db892096421d1a9f48f3
 * Jorn Vernee ([jvernee](@JornVernee) - Committer) ⚠️ Review applies to 38c81345427b02034d063d824ab6f982434f544a
 * Paul Sandoz ([psandoz](@PaulSandoz) - Committer) ⚠️ Review applies to 86c572f3281a16f5fa405f8e8027b71158d5f482
 * Henry Jen ([henryjen](@slowhog) - Committer) ⚠️ Review applies to 86c572f3281a16f5fa405f8e8027b71158d5f482


### Download
`$ git fetch https://git.openjdk.java.net/panama-foreign pull/222/head:pull/222`
`$ git checkout pull/222`

sundararajana commented on src/jdk.incubator.foreign/share/classes/jdk/incubator/foreign/NativeScope.java:
```diff
@@ -49,6 +52,8 @@
 public abstract class NativeScope implements AutoCloseable {
 
+    Unsafe unsafe = Unsafe.getUnsafe();
+
```
should this field be private?

sundararajana replied:
looks like unused. may be removed

JornVernee commented on src/jdk.incubator.foreign/share/classes/jdk/incubator/foreign/NativeScope.java:
```diff
@@ -176,31 +181,167 @@ public MemoryAddress allocate(MemoryLayout layout, double value) {
-        MemoryAddress addr = allocate(layout);
-        handle.set(addr, value);
+    public MemoryAddress allocateArray(ValueLayout elementLayout, byte[] array) {
+        if (elementLayout.byteSize() != 1) {
+            throw new IllegalArgumentException("Bad layout size");
+        }
```
Can maybe be changed to 
```suggestion
        Utils.checkPrimitiveCarrierCompat(byte.class, elementLayout);
```

JornVernee commented on src/jdk.incubator.foreign/share/classes/jdk/incubator/foreign/NativeScope.java:
```diff
@@ -176,31 +181,167 @@ public MemoryAddress allocate(MemoryLayout layout, double value) {
+            throw new IllegalArgumentException("Bad layout size");
+        }
+        MemoryAddress addr = allocate(MemoryLayout.ofSequence(array.length, elementLayout));
+        if (elementLayout.order() == ByteOrder.nativeOrder()) {
+            addr.segment().copyFrom(MemorySegment.ofArray(array));
+        } else {
+            ((AbstractMemorySegmentImpl)addr.segment()).copyFromSwap(MemorySegment.ofArray(array), elementLayout.byteSize());
+        }
```
This code seems to be the same in the other implementations (except for the call to MS::ofArray). Maybe you want to factor it out into a helper method?

JornVernee commented on src/jdk.incubator.foreign/share/classes/jdk/internal/foreign/AbstractMemorySegmentImpl.java:
```diff
@@ -132,6 +132,16 @@ public void copyFrom(MemorySegment src) {
                 base(), min(), size);
     }
 
+    public void copyFromSwap(MemorySegment src, long elemSize) {
```
Should this maybe be a public API in MemorySegment as well?

mcimadamore replied:
> Should this maybe be a public API in MemorySegment as well?

I'll prefer to leave that discussion for another patch

JornVernee replied:
Ok. Was just wondering if you had any thoughts about that already :)

mcimadamore commented:
I've addressed most of the comments; I also added an extra method `allocateArray(MemoryLayout, long)` to allocate an 'empty' array with given layout. This "just" a shortcut for `allocate(MemoryLayout.ofSequence(...))` but I think it's very handy - e.g. to create empty pointer arrays, as in:

```java
MemoryAddress addr = scope.allocateArray(C_POINTER, 42);
```

JornVernee commented on src/jdk.incubator.foreign/share/classes/jdk/incubator/foreign/NativeScope.java:
```diff
@@ -356,6 +354,25 @@ public MemoryAddress allocate(MemoryLayout layout) {
+    allocate(MemoryLayout.ofSequence(size, elementLayout));
+     * }</pre>
+     * @param elementLayout the array element layout.
+     * @param size the array size.
```
Maybe useful to mention that the size is the element count, and not in bytes.

JornVernee commented on src/jdk.incubator.foreign/share/classes/jdk/incubator/foreign/NativeScope.java:
```diff
@@ -356,6 +354,25 @@ public MemoryAddress allocate(MemoryLayout layout) {
+     * @return an address which points to the newly allocated memory block.
+     * @throws OutOfMemoryError if there is not enough space left in this native scope, that is, if
+     * {@code limit() - size() < (elementLayout.byteSize() * size)}.
+     * @throws IllegalArgumentException if {@code elementLayout.byteSize()} does not conform to the size of a double value.
```
This looks like a remnant of copy-pasting?

mcimadamore commented:
/integrate
